### PR TITLE
gRPC Zap ReplaceGrpcLoggerV2

### PR DESCRIPTION
**Pull request by kush-patel-hs**
We at Hootsuite were noticing noisy logs coming from grpc and we noticed it's because grpc_zap implements logger v1 interface. We need the logger v2 implemented instead, so I have opened this PR.

This is a revived and comments addressed version of #144 which was opened by @jamisonhyatt. Feel free to just merge this in after approval if it looks good!

### CODE REVIEW

File: logging/zap/examples_test.go [hunk 1]
```diff
@@ -4,10 +4,10 @@ import (
...
 	"time"
 
-	"github.com/grpc-ecosystem/go-grpc-middleware"
-	"github.com/grpc-ecosystem/go-grpc-middleware/logging/zap"
-	"github.com/grpc-ecosystem/go-grpc-middleware/tags"
-	"github.com/grpc-ecosystem/go-grpc-middleware/tags/zap"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	grpc_zap "github.com/grpc-ecosystem/go-grpc-middleware/logging/zap"
+	grpc_ctxtags "github.com/grpc-ecosystem/go-grpc-middleware/tags"
+	ctx_zap "github.com/grpc-ecosystem/go-grpc-middleware/tags/zap"
```
Comment (kush-patel-hs) [inline logging/zap/examples_test.go]: newer version of goimports does this because the package name is not the same as the last value in the path, let me know if you'll like me to undo this change

**Comment by kush-patel-hs** [comment]
Other path would be instead of introducing ReplaceGrpcLoggerV2, just change ReplaceGrpcLogger internals. The interface to the user doesn't change (input is still only a zap.Logger).

**Comment by kush-patel-hs** [comment]
I signed it!

**Comment by kush-patel-hs** [comment]
@domgreen

**Comment by kush-patel-hs** [comment]
@devnev

File: logging/zap/grpclogger.go [hunk 2]
```diff
@@ -44,3 +47,103 @@ func (l *zapGrpcLogger) Printf(format string, args ...interface{}) {
 func (l *zapGrpcLogger) Println(args ...interface{}) {
 	l.logger.Info(fmt.Sprint(args...))
 }
+
+// ReplaceGrpcLoggerV2 replaces the grpc_log.LoggerV2 with the input zap.Logger
+// This logger adheres to the grpc go environment variables GRPC_GO_LOG_VERBOSITY_LEVEL and GRPC_GO_LOG_SEVERITY_LEVEL.
```
Comment (devnev) [inline logging/zap/grpclogger.go]: What is the use case for adding the extra log severity level handling instead of accepting the log level of the zap.Logger as-is?

Comment (kush-patel-hs) [reply logging/zap/grpclogger.go]: If I want my application to log at the info level and want the grpc system to log at error using the same global logger instance. Plus matches https://github.com/grpc/grpc-go/blob/24b2fb8959201be9ce659bc87b0d590a34c67eae/grpclog/loggerv2.go#L124. Thoughts? I can remove the severity level if you like and we could add it in the future if we feel the need.

Comment (devnev) [reply logging/zap/grpclogger.go]: I would favour starting with the explicit version: use the incoming logger's severity - and for verbosity, accept an argument with the desired verbosity level.

Comment (kush-patel-hs) [reply logging/zap/grpclogger.go]: What's your thoughts on having an option to allow for setting verbosity level and defaulting to 0?

Comment (kush-patel-hs) [reply logging/zap/grpclogger.go]: Updated @devnev

File: logging/zap/examples_test.go (hunk 1 above)
Comment (domgreen) [reply logging/zap/examples_test.go]: This needs to be fixed at some point not for this PR :+1:

File: logging/zap/grpclogger.go [hunk 3]
```diff
@@ -44,3 +45,80 @@ func (l *zapGrpcLogger) Printf(format string, args ...interface{}) {
 func (l *zapGrpcLogger) Println(args ...interface{}) {
 	l.logger.Info(fmt.Sprint(args...))
 }
+
+// GrpcLoggerV2Options allows you to set grpclogger options when using ReplaceGrpcLoggerV2
+type GrpcLoggerV2Options struct {
+	verbosity int
+}
+
+// DefaultGrpcLoggerV2Options is a set of sane default options for ReplaceGrpcLoggerV2
```
Comment (domgreen) [inline logging/zap/grpclogger.go]: nit - would remove `sane` and add a `.` at the end of the sentence.

Comment (kush-patel-hs) [reply logging/zap/grpclogger.go]: i added `.` in all my exported struct/function comments :)

**Comment by kush-patel-hs** [comment]
Thanks @domgreen! Pretty excited to get this in, will help us clear up some noisy logs from using the v1 logger at Hootsuite

File: logging/zap/grpclogger.go [hunk 4]
```diff
@@ -44,3 +45,80 @@ func (l *zapGrpcLogger) Printf(format string, args ...interface{}) {
 func (l *zapGrpcLogger) Println(args ...interface{}) {
 	l.logger.Info(fmt.Sprint(args...))
 }
+
+// GrpcLoggerV2Options allows you to set grpclogger options when using ReplaceGrpcLoggerV2.
+type GrpcLoggerV2Options struct {
+	verbosity int
```
Comment (domgreen) [inline logging/zap/grpclogger.go]: This is not accessible :( you would never be able to set this.

File: logging/zap/grpclogger.go [hunk 5]
```diff
@@ -44,3 +45,80 @@ func (l *zapGrpcLogger) Printf(format string, args ...interface{}) {
...
+}
+
+// DefaultGrpcLoggerV2Options is a set of default options for ReplaceGrpcLoggerV2.
+var DefaultGrpcLoggerV2Options = &GrpcLoggerV2Options{}
+
+// ReplaceGrpcLoggerV2 replaces the grpc_log.LoggerV2 with the input zap.Logger.
+// This logger allows you to set grpc verbosity level.
+func ReplaceGrpcLoggerV2(logger *zap.Logger, opts *GrpcLoggerV2Options) {
```
Comment (domgreen) [inline logging/zap/grpclogger.go]: Lets change this to `ReplaceLoggerV2(logger *zap.Logger)` this will match the upstream here https://github.com/grpc/grpc-go/blob/master/grpclog/loggerv2.go#L101

Comment (domgreen) [reply logging/zap/grpclogger.go]: Add a second `ReplaceLoggerV2WithVerbosity` with the option to pass in `v int` as https://github.com/grpc/grpc-go/blob/master/grpclog/loggerv2.go#L107 again matching upstream

Comment (domgreen) [reply logging/zap/grpclogger.go]: Change `ReplaceLoggerV2` to call `ReplaceLoggerV2WithVerbosity(logger , 0)` see https://github.com/grpc/grpc-go/blob/master/grpclog/loggerv2.go#L102

File: logging/zap/grpclogger.go [hunk 6]
```diff
@@ -44,3 +45,80 @@ func (l *zapGrpcLogger) Printf(format string, args ...interface{}) {
...
 	l.logger.Info(fmt.Sprint(args...))
 }
+
+// GrpcLoggerV2Options allows you to set grpclogger options when using ReplaceGrpcLoggerV2.
+type GrpcLoggerV2Options struct {
+	verbosity int
+}
+
+// DefaultGrpcLoggerV2Options is a set of default options for ReplaceGrpcLoggerV2.
+var DefaultGrpcLoggerV2Options = &GrpcLoggerV2Options{}
```
Comment (domgreen) [inline logging/zap/grpclogger.go]: remove this and GrpcLoggerV2Options once done the below comments :+1:

File: logging/zap/grpclogger.go [hunk 7]
```diff
@@ -44,3 +45,80 @@ func (l *zapGrpcLogger) Printf(format string, args ...interface{}) {
...
+
+// ReplaceGrpcLoggerV2 replaces the grpc_log.LoggerV2 with the input zap.Logger.
+// This logger allows you to set grpc verbosity level.
+func ReplaceGrpcLoggerV2(logger *zap.Logger, opts *GrpcLoggerV2Options) {
+	zgl := &zapGrpcLoggerV2{Logger: logger.With(SystemField, zap.Bool("grpc_log", true))}
```
Comment (domgreen) [inline logging/zap/grpclogger.go]: Inside Relace With Verbosity you can then do :
```
zgl := &zapGrpcLoggerV2{
  Logger: logger.With(SystemField, zap.Bool("grpc_log", true)),
  verbosity: 0,
}
```

File: logging/zap/grpclogger.go (hunk 5 above)
Comment (kush-patel-hs) [reply logging/zap/grpclogger.go]: 🤔 I can do that! Want to drop the Grpc from the name?

File: logging/zap/grpclogger.go (hunk 4 above)
Comment (kush-patel-hs) [reply logging/zap/grpclogger.go]: Oh whoops lol :(

File: logging/zap/grpclogger.go (hunk 7 above)
Comment (kush-patel-hs) [reply logging/zap/grpclogger.go]: done

File: logging/zap/grpclogger.go (hunk 6 above)
Comment (kush-patel-hs) [reply logging/zap/grpclogger.go]: done

File: logging/zap/grpclogger.go [hunk 8]
```diff
@@ -44,3 +45,75 @@ func (l *zapGrpcLogger) Printf(format string, args ...interface{}) {
 func (l *zapGrpcLogger) Println(args ...interface{}) {
 	l.logger.Info(fmt.Sprint(args...))
 }
+
+// ReplaceGrpcLoggerV2 replaces the grpc_log.LoggerV2 with the input zap.Logger.
+// This logger allows you to set grpc verbosity level.
```
Comment (domgreen) [inline logging/zap/grpclogger.go]: This is only the WithVerbosity one

File: logging/zap/grpclogger.go [hunk 9]
```diff
@@ -44,3 +45,75 @@ func (l *zapGrpcLogger) Printf(format string, args ...interface{}) {
...
+func ReplaceGrpcLoggerV2(logger *zap.Logger) {
+	ReplaceGrpcLoggerV2WithVerosity(logger, 0)
+}
+
+// ReplaceGrpcLoggerV2WithVerosity replaces the grpc_.LoggerV2 with the provided logger and verbosity.
+func ReplaceGrpcLoggerV2WithVerosity(logger *zap.Logger, verbosity int) {
```
Comment (domgreen) [inline logging/zap/grpclogger.go]: Verosity > Verbosity

Comment (kush-patel-hs) [reply logging/zap/grpclogger.go]: done

File: logging/zap/grpclogger.go (hunk 8 above)
Comment (kush-patel-hs) [reply logging/zap/grpclogger.go]: true, done

**Comment by kush-patel-hs** [comment]
Thanks for all the time you two spent on reviewing and providing feedback so far @domgreen @devnev :) I think I've addressed all feedbacks now, let's push this through :)

**Comment by kush-patel-hs** [comment]
Thanks for all the help 👍

**Comment by fproulx-eoscanada** [comment]
@kush-patel-hs Hi! I tried this. My default `zap.Logger` instance is set to `Info` level and it appears that setting the severity to `2` like `grpc_zap.ReplaceGrpcLoggerV2WithVerbosity(zlog, 2)` does not lower the noise. Skimming through the code this `severity` appears ? to be ignored ? So should I create a different `zap.Logger` with `Error` level before calling `ReplaceGrpcLoggerV2WithVerbosity` ?

**Comment by kush-patel-hs** [comment]
@fproulx-eoscanada verbosity `2` will increase noise, the higher the verbosity number the more noise (i.e max noise is 99). Less noise is 0. Verbosity will be set to 0 if you simply do `grpc_zap.ReplaceGrpcLoggerV2(zlog)`. You would notice the LoopyWriter logs transport closing in particular go away changing from verbosity 2 to verbosity 0. If you want even less noise than that than yes create a new zap logger with Error or Warn level before doing the replace.

**Comment by fproulx-eoscanada** [comment]
@kush-patel-hs ah! great so let's set it back to `0` and try again =)
Thanks -- great contribution by Hootsuite. We love grpc and zap too =)

**Comment by kush-patel-hs** [comment]
Awesome :) let me know if you need any more help! 
And thank you, appreciate it :)

**Comment by Icedroid** [comment]
Who can add logrus log with ReplaceGrpcLoggerV2?

**Comment by srp** [comment]
In case others run across this thread, you might want or need to pass a sublogger with a different level then your root logger. This can look something like:

```
grpc_zap.ReplaceGrpcLoggerV2(logger.WithOptions(zap.IncreaseLevel(zap.WarnLevel))
```
